src/AssociatedJobInfoClient.cs

```diff
@@ -0,0 +1,124 @@
+using Microsoft.DotNet.HelixPoolProvider.Models;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.HelixPoolProvider
+{
+    public class AssociatedJobInfo
+    {
+        public string BuildSourceBranch { get; }
+        public string SystemPullRequestTargetBranch { get; }
+
+        public static readonly AssociatedJobInfo Empty
+            = new AssociatedJobInfo("", "");
+
+        public AssociatedJobInfo(
+            string buildSourceBranch,
+            string systemPullRequestTargetBranch)
+        {
+            BuildSourceBranch = buildSourceBranch;
+            SystemPullRequestTargetBranch = systemPullRequestTargetBranch;
+        }
+    }
+
+    public class AssociatedJobInfoClient
+    {
+        private readonly IHttpClientFactory _httpClientFactory;
+        private readonly ILogger<AssociatedJobInfoClient> _logger;
+
+        public AssociatedJobInfoClient(
+            IHttpClientFactory httpClientFactory,
+            ILogger<AssociatedJobInfoClient> logger)
+        {
+            _httpClientFactory = httpClientFactory;
+            _logger = logger;
+        }
+
+        public async Task<AssociatedJobInfo> TryGetAssociatedJobInfo(
+            string getAssociatedJobUrl,
+            string authenticationToken)
+        {
+            try
+            {
+                var agentRequestJob = await TryGetAssociatedAgentRequestJob(
+                    getAssociatedJobUrl,
+                    authenticationToken);
+
+                if (agentRequestJob != null)
+                {
+                    return ParseAssociatedJobInfo(agentRequestJob);
+                }
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(
+                    exception,
+                    "Unable to get associated job info from {getAssociatedJobUrl} because of an exception",
+                    getAssociatedJobUrl);
+            }
+
+            return AssociatedJobInfo.Empty;
+        }
+
+        private async Task<AgentRequestJob> TryGetAssociatedAgentRequestJob(
+            string getAssociatedJobUrl,
+            string authenticationToken)
+        {
+            if (!string.IsNullOrEmpty(getAssociatedJobUrl))
+            {
+                _logger.LogInformation("Getting associated job from {AssociatedJobUrl}", getAssociatedJobUrl);
+
+                var httpClient = _httpClientFactory.CreateClient();
+                var message = new HttpRequestMessage(HttpMethod.Get, getAssociatedJobUrl);
```

Comment:
We need to lower the timeout on these requests. It is _critical_ that the pool provider responds in <30 seconds.  AzDO can be very flaky, so we can't afford the default timeout of 100s, since that will cause the build to get dropped on the floor.

Reply:
I have set the timeout to 10 seconds.